File: java/mlt-core/src/main/java/org/maplibre/mlt/metadata/stream/StreamMetadata.java
package org.maplibre.mlt.metadata.stream;

import com.google.common.primitives.Bytes;
import java.io.IOException;
import me.lemire.integercompression.IntWrapper;
import org.maplibre.mlt.converter.encodings.EncodingUtils;
import org.maplibre.mlt.decoder.DecodingUtils;

public class StreamMetadata {
  private final PhysicalStreamType physicalStreamType;
  private final LogicalStreamType logicalStreamType;
  private final LogicalLevelTechnique logicalLevelTechnique1;
  private final LogicalLevelTechnique logicalLevelTechnique2;
  private final PhysicalLevelTechnique physicalLevelTechnique;
  /* After logical Level technique was applied -> when rle is used it is the length of the runs and values array */
  private final int numValues;
  private final int byteLength;

  // TODO: refactor -> use builder pattern
  public StreamMetadata(
      PhysicalStreamType physicalStreamType,
      LogicalStreamType logicalStreamType,
      LogicalLevelTechnique logicalLevelTechnique1,
      LogicalLevelTechnique logicalLevelTechnique2,
      PhysicalLevelTechnique physicalLevelTechnique,
      int numValues,
      int byteLength) {
    this.physicalStreamType = physicalStreamType;
    this.logicalStreamType = logicalStreamType;
    this.logicalLevelTechnique1 = logicalLevelTechnique1;
    this.logicalLevelTechnique2 = logicalLevelTechnique2;
    this.physicalLevelTechnique = physicalLevelTechnique;
    this.numValues = numValues;
    this.byteLength = byteLength;
  }

  private int getLogicalType() {
    if (logicalStreamType == null) {
      return 0;
    }

    if (logicalStreamType.dictionaryType() != null) {
      return logicalStreamType.dictionaryType().ordinal();
    }

    if (logicalStreamType.lengthType() != null) {
      return logicalStreamType.lengthType().ordinal();
    }

    return logicalStreamType.offsetType().ordinal();
  }

  public byte[] encode() throws IOException {
    final var encodedStreamType = (byte) ((physicalStreamType.ordinal()) << 4 | getLogicalType());
    final var encodedEncodingScheme =
        (byte)
            (logicalLevelTechnique1.ordinal() << 5
                | logicalLevelTechnique2.ordinal() << 2
                | physicalLevelTechnique.ordinal());
    final var encodedLengthInfo =
        EncodingUtils.encodeVarints(new int[] {numValues, byteLength}, false, false);
    return Bytes.concat(new byte[] {encodedStreamType, encodedEncodingScheme}, encodedLengthInfo);
  }

  public static StreamMetadata decode(byte[] tile, IntWrapper offset) throws IOException {
    var streamType = tile[offset.get()];
    var physicalStreamType = PhysicalStreamType.values()[streamType >> 4];
    LogicalStreamType logicalStreamType =
        switch (physicalStreamType) {
          case DATA -> new LogicalStreamType(DictionaryType.values()[streamType & 0xf]);
          case OFFSET -> new LogicalStreamType(OffsetType.values()[streamType & 0xf]);
          case LENGTH -> new LogicalStreamType(LengthType.values()[streamType & 0xf]);
          default -> null;
        };
    offset.increment();

    var encodingsHeader = tile[offset.get()] & 0xFF;
    var logicalLevelTechnique1 = LogicalLevelTechnique.values()[encodingsHeader >> 5];
    var logicalLevelTechnique2 = LogicalLevelTechnique.values()[encodingsHeader >> 2 & 0x7];
    var physicalLevelTechnique = PhysicalLevelTechnique.values()[encodingsHeader & 0x3];
    offset.increment();
    var sizeInfo = DecodingUtils.decodeVarints(tile, offset, 2);
    var numValues = sizeInfo[0];
    var byteLength = sizeInfo[1];

    return new StreamMetadata(
        physicalStreamType,
        logicalStreamType,
        logicalLevelTechnique1,
        logicalLevelTechnique2,
        physicalLevelTechnique,
        numValues,
        byteLength);
  }

  public PhysicalStreamType physicalStreamType() {
    return this.physicalStreamType;
  }

  public LogicalStreamType logicalStreamType() {
    return this.logicalStreamType;
  }

  public LogicalLevelTechnique logicalLevelTechnique1() {
    return this.logicalLevelTechnique1;
  }

  public LogicalLevelTechnique logicalLevelTechnique2() {
    return this.logicalLevelTechnique2;
  }

  public PhysicalLevelTechnique physicalLevelTechnique() {
    return this.physicalLevelTechnique;
  }

  public int numValues() {
    return this.numValues;
  }

  public int byteLength() {
    return this.byteLength;
  }
}

File: ts/src/metadata/tile/streamMetadata.ts
import { PhysicalStreamType } from "./physicalStreamType";
import { LogicalStreamType } from "./logicalStreamType";
import { LogicalLevelTechnique } from "./logicalLevelTechnique";
import { PhysicalLevelTechnique } from "./physicalLevelTechnique";
import { DictionaryType } from "./dictionaryType";
import { OffsetType } from "./offsetType";
import { LengthType } from "./lengthType";
import { decodeVarintInt32 } from "../../decoding/integerDecodingUtils";
import type IntWrapper from "../../decoding/intWrapper";

export function decodeStreamMetadata(tile: Uint8Array, offset: IntWrapper): StreamMetadata {
    const stream_type = tile[offset.get()];
    const physicalStreamType = Object.values(PhysicalStreamType)[stream_type >> 4] as PhysicalStreamType;
    let logicalStreamType: LogicalStreamType | null = null;

    switch (physicalStreamType) {
        case PhysicalStreamType.DATA:
            logicalStreamType = new LogicalStreamType(
                Object.values(DictionaryType)[stream_type & 0xf] as DictionaryType,
            );
            break;
        case PhysicalStreamType.OFFSET:
            logicalStreamType = new LogicalStreamType(null, Object.values(OffsetType)[stream_type & 0xf] as OffsetType);
            break;
        case PhysicalStreamType.LENGTH:
            logicalStreamType = new LogicalStreamType(
                null,
                null,
                Object.values(LengthType)[stream_type & 0xf] as LengthType,
            );
            break;
    }
    offset.increment();

    const encodings_header = tile[offset.get()];
    const llt1 = Object.values(LogicalLevelTechnique)[encodings_header >> 5] as LogicalLevelTechnique;
    const llt2 = Object.values(LogicalLevelTechnique)[(encodings_header >> 2) & 0x7] as LogicalLevelTechnique;
    const plt = Object.values(PhysicalLevelTechnique)[encodings_header & 0x3] as PhysicalLevelTechnique;
    offset.increment();

    const sizeInfo = decodeVarintInt32(tile, offset, 2);
    const numValues = sizeInfo[0];
    const byteLength = sizeInfo[1];

    return {
        physicalStreamType,
        logicalStreamType,
        logicalLevelTechnique1: llt1,
        logicalLevelTechnique2: llt2,
        physicalLevelTechnique: plt,
        numValues,
        byteLength,
        decompressedCount: numValues,
    };
}

export type StreamMetadata = {
    readonly physicalStreamType: PhysicalStreamType;
    readonly logicalStreamType: LogicalStreamType;
    readonly logicalLevelTechnique1: LogicalLevelTechnique;
    readonly logicalLevelTechnique2: LogicalLevelTechnique;
    readonly physicalLevelTechnique: PhysicalLevelTechnique;
    readonly numValues: number;
    readonly byteLength: number;
    /**
     * Returns the number of decompressed values.
     * For non-RLE streams, this is the same as numValues.
     * For RLE streams, this is overridden to return numRleValues.
     */
    readonly decompressedCount: number;
};
